File: 0/0/7/007.c
// Copyright (C) 2013 Jorge Aparicio

#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>

static int size = 100000;
static int target = 10000;

int main() {
  int i, count = 1;
  long j, p;
  bool *sieve = (bool *)calloc(size, sizeof(bool));

  for (i = 0; i < size; i++) {
    if (!sieve[i]) {
      p = 2*i + 3;

      if (count == target) {
        printf("%d\n", p);
        break;
      }

      for (j = p*p; j < 2*size + 3; j += 2*p)
        sieve[(j - 3) / 2] = true;

      count++;
    }
  }

  free(sieve);

  return 0;
}

File: 0/0/7/007.rs
// Copyright (C) 2014 Jorge Aparicio

use std::iter::range_step;

static size: int = 100000;
static target: int = 10000;

fn main() {
  let mut sieve: [bool, ..size] = [true, ..size];
  let mut count = 1;

  for i in range(0, size) {
    if sieve[i] {
      let p = 2*i + 3;

      if count == target {
        println!("{}", p);
        break;
      }

      for j in range_step(p * p, 2*size + 3, 2*p) {
        sieve[(j - 3) / 2] = false;
      }

      count += 1;
    }
  }
}

